keps/sig-node/3960-pod-lifecycle-sleep-action/README.md

```diff
@@ -0,0 +1,363 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes the addition of a new sleep action for the PreStop lifecycle hook in Kubernetes, allowing containers to pause for a specified duration before termination. This enhancement aims to provide a more straightforward way to manage graceful shutdowns and improve the overall lifecycle management of containers.
+
+## Motivation
+
+Currently, Kubernetes supports two types of actions for PreStop hooks: exec and httpGet. Although these actions offer flexibility, they often require additional scripting or custom solutions to achieve a simple sleep functionality. A built-in sleep action would provide a more user-friendly and native solution for scenarios where a container needs to pause before shutting down, such as:
+
+- Ensuring that the container gracefully releases resources and connections.
+- Allowing a smooth transition in load balancers or service meshes.
+- Providing a buffer period for external monitoring and alerting systems.
+
+### Goals
+
+- Allow containers to perform cleanup or shutdown actions before being terminated, by sleeping for a specified duration in the preStop hook.
+- Improve the overall reliability and availability of Kubernetes applications by providing a way for containers to gracefully terminate.
+
+### Non-Goals
+
+- This KEP does not aim to replace other Kubernetes features that can be used to perform cleanup actions, such as init containers or sidecar containers.
+- This KEP does not aim to provide a way to pause or delay pod termination indefinitely.
+
+## Proposal
+
+We propose adding a new sleep action for the PreStop hook, which will pause the container for a specified duration before termination. The API changes will include the following:
+
+- Extending the LifecycleHandler object to support a new Sleep field.
+- Adding a SleepAction object that includes a Duration field to specify the sleep period in seconds.
+
+### User Stories (Optional)
+
+#### Story 1
+As a Kubernetes user, I want to configure my container to sleep for a specific duration during terminating with grace.And I want to do it without needing a sleep binary in my image.
+
+### Risks and Mitigations
+
+One potential risk is that if the sleep duration is longer than terminationGracePeriodSeconds, container will be killed before sleep action finishes, and can cause unpredictable consequences.
```

Comment:
I am not following on component will experience the `unpredictable consequences`. For kubelet it should be almost a normal situation as there are so many cases when this will happen.

Reply:
Agree - early termination is normal.